bundles/org.openhab.binding.adorne/src/main/java/org/openhab/binding/adorne/internal/hub/AdorneHubController.java

```diff
@@ -0,0 +1,546 @@
+        CompletableFuture<List<Integer>> zoneCommand = new CompletableFuture<>();
+        synchronized (this) {
+            zoneCommands.add(zoneCommand);
+        }
+        sendRestCmd(String.format(HUB_REST_REQUEST_ZONES, getNextCommandId()));
+        return zoneCommand;
+    }
+
+    /**
+     * Gets asynchronously the MAC address of the hub.
+     *
+     * @return a future for the MAC address
+     */
+    public CompletableFuture<String> getMACAddress() {
+        CompletableFuture<String> macAddressCommand = new CompletableFuture<>();
+        synchronized (this) {
+            macAddressCommands.add(macAddressCommand);
+        }
+        sendRestCmd(String.format(HUB_REST_REQUEST_MACADDRESS, getNextCommandId()));
+        return macAddressCommand;
+    }
+
+    private void sendRestCmd(String cmd) {
+        logger.debug("Sending command {}", cmd);
+        synchronized (this) {
+            PrintStream hubOut = this.hubOut;
+            if (hubOut != null) {
+                hubOut.print(cmd);
+            } else {
+                throw new IllegalStateException("Can't send command. Adorne Hub connection is not available.");
+            }
+        }
+    }
+
+    /**
+     * Sets a listener to be notified on hub changes.
+     *
+     * @param changeListener instance to be notified
+     */
+    public synchronized void setChangeListener(AdorneHubChangeNotify changeListener) {
+        this.changeListener = changeListener;
+    }
+
+    /**
+     * Runs the controller that is interacting with the Adorne Hub by sending commands and listening for updates
+     */
+    @Override
+    public void run() {
+        logger.info("Starting hub controller");
+        try {
+            Map<String, @Nullable Object> hubMsg;
+            String service;
+            hubReconnectSleep = HUB_RECONNECT_SLEEP_MIN;
+
+            // Main message loop listening for updates from the hub
+            while (!shouldStop()) {
+                if (!connect()) {
+                    continue;
+                }
+
+                if ((hubMsg = getHubMsg()) == null) {
+                    continue;
+                }
+
+                // Process message based on service type
+                if ((service = (String) hubMsg.get(HUB_TOKEN_SERVICE)) == null) {
+                    continue; // Ignore messages that don't have a service specified
+                }
+
+                if (service.equals(HUB_SERVICE_REPORT_ZONE_PROPERTIES)) {
+                    processMsgReportZoneProperties(hubMsg);
+                } else if (service.equals(HUB_SERVICE_ZONE_PROPERTIES_CHANGED)) {
+                    processMsgZonePropertiesChanged(hubMsg);
+                } else if (service.equals(HUB_SERVICE_LIST_ZONE)) {
+                    processMsgListZone(hubMsg);
+                } else if (service.equals(HUB_SERVICE_SYSTEM_INFO)) {
+                    processMsgSystemInfo(hubMsg);
+                }
+            }
+        } catch (Exception e) {
```

Comment:
Please do not catch `Exception`, catch specific exceptions that can apply here. (also applies below)

Reply:
All specific exceptions are caught in the respective methods in the loop. This is truly a generic error handler - any exception coming through is fatal, disconnects the hub and is logged as Error. Isn't that the desired behavior over letting them bubble up further?

Reply:
Then it would be `RuntimeException` instead of `Exception`

Reply:
Good point. Done.